### PR TITLE
bug: invalid CSOD related course keys are now encoded

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
[jira](https://openedx.atlassian.net/browse/ENT-4663)

CSOD customers cannot accept courses that contain keys with invalid chars- thus for any **invalid course key**  we export, encoded it. When exporting learner data, make sure to encode the course ID if it's invalid as well in order to find the right course.

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: integrated_channels/cornerstone/client.py [hunk 1]
```diff
@@ -94,6 +101,17 @@ def create_course_completion(self, user_id, payload):  # pylint: disable=unused-
         )
         return response.status_code, response.text
 
+    def _convert_invalid_course_ids(self, course_id):
+        """
+        Regex check a course ID to see if it contains any invalid chars. If it does then encode the string, otherwise
+        return the original course ID.
+        """
+        re2 = re.compile(r"[|<>.&%\s\\/\“]+")
+        if re2.search(course_id):
```
Comment (mattdrayer) [inline integrated_channels/cornerstone/client.py:110]: I guess I feel like, if we are converting course keys to send to CSOD, we should just convert all of the keys since it might look weird/buggy if some keys are in the traditional format and some keys are in the encoded format.  Thoughts?

File: integrated_channels/cornerstone/exporters/content_metadata.py [hunk 2]
```diff
@@ -53,6 +55,32 @@ def export_force_all_catalogs(self):
         """
         return self.export(force_retrieve_all_catalogs=True)
 
+    def encode_invalid_keys(self, content_key):
```
Comment (mattdrayer) [inline integrated_channels/cornerstone/exporters/content_metadata.py:58]: Why is this here when there is something similar in `client.py` above?  This seems like more of a utility function -- maybe it should be moved to `cornerstone/utils.py`

File: tests/test_integrated_channels/test_cornerstone/test_exporters/test_content_metadata.py [hunk 3]
```diff
@@ -113,6 +114,50 @@ def test_transform_description(self, item_description, expected_description):
...
+            encode_course_key_into_base64('edx+|'),
+        ),
+        (
+            {'key': 'edx+/'},
+            encode_course_key_into_base64('edx+/'),
+        ),
+    )
+    @responses.activate
+    @ddt.unpack
+    def test_transform_ke (self, item_key, expected_id):
```
Comment (edx-abolger) [inline tests/test_integrated_channels/test_cornerstone/test_exporters/test_content_metadata.py:153]: Please add a test case where invalid characters appear in the beginning and middle of the keys, not just at the end.

File: integrated_channels/cornerstone/client.py (hunk 1 above)
Comment (edx-abolger) [reply integrated_channels/cornerstone/client.py:110]: The problem is that regenerating course keys has far reaching effects since I believe those keys are used as IDs on the Cornerstone side. It is possible any dropped and recreated courses would lose linked learner data and enrollments on the Cornerstone side. At a minimum, there would suddenly be two of every course and we'd have to handle that with additional manual clean up code. This solution is for backwards compatibility. (A comment on this in code would not be amiss, @alex-sheehan-edx )

**Review by edx-abolger — CHANGES_REQUESTED** [review]
Small requests, should be easily resolvable

File: integrated_channels/cornerstone/client.py (hunk 1 above)
Comment (mattdrayer) [reply integrated_channels/cornerstone/client.py:110]: Makes perfect sense -- I neglected to consider the fact that there are already edX courses in CSOD tenants and we would have effectively abandoned those records on the CSOD side.

File: integrated_channels/cornerstone/exporters/content_metadata.py (hunk 2 above)
Comment (alex-sheehan-edx) [reply integrated_channels/cornerstone/exporters/content_metadata.py:58]: good point! Yeah, this should really be in a utils file. 👍

**Review by edx-abolger — APPROVED** [review]
